controlpanel/api/cluster.py

```diff
@@ -463,121 +463,139 @@ class ToolDeploymentError(Exception):
+            "aws.iamRole": self.user.iam_role_name,
+            "toolsDomain": settings.TOOLS_DOMAIN,
+        }
 
-    if len(deployments) > 1:
-        log.warning(f"Multiple matches for {tool_deployment!r} found")
-        raise MultipleObjectsReturned(tool_deployment)
+        # generate per-deployment secrets
+        for key, value in self.tool.values.items():
+            if value == "<SECRET_TOKEN>":
+                self.tool.values[key] = secrets.token_hex(32)
 
-    return deployments[0]
+        values.update(self.tool.values)
+        values.update(kwargs)
+        set_values = []
+        for key, val in values.items():
+            escaped_val = val.replace(',', '\,')
+            set_values.extend(['--set', f'{key}={escaped_val}'])
 
+        try:
+            return helm.upgrade_release(
+                self.release_name,
+                f'{settings.HELM_REPO}/{self.chart_name}',  # XXX assumes repo name
+                # f'--version', tool.version,
+                f'--namespace', self.k8s_namespace,
+                *set_values,
+            )
+
+        except HelmError as error:
+            raise ToolDeploymentError(error)
+
+    def uninstall(self, id_token):
+        deployment = self.get_deployment(id_token)
+        helm.delete(
+            deployment.metadata.name,
+            f"--namespace={self.k8s_namespace}",
+        )
 
-def delete_tool_deployment(tool_deployment, id_token):
-    deployment = get_tool_deployment(tool_deployment, id_token)
-    helm.delete(
-        deployment.metadata.name,
-        f"--namespace={tool_deployment.user.k8s_namespace}",
-    )
+    def restart(self, id_token):
+        k8s = KubernetesClient(id_token=id_token)
+        return k8s.AppsV1Api.delete_collection_namespaced_replica_set(
+            self.k8s_namespace,
+            label_selector=(
+                f"app={self.chart_name}"
+                # f'-{tool_deployment.tool.version}'
+            ),
+        )
 
+    @classmethod
+    def get_deployments(cls, user, id_token, search_name=None, search_version=None):
+        deployments = []
+        k8s = KubernetesClient(id_token=id_token)
+        results = k8s.AppsV1Api.list_namespaced_deployment(user.k8s_namespace)
+        for deployment in results.items:
+            app_name = deployment.metadata.labels["app"]
+            _, version = deployment.metadata.labels["chart"].rsplit("-", 1)
+            if search_name and search_name not in app_name:
+                continue
+            if search_version and not version.startswith(search_version):
+                continue
+            deployments.append(deployment)
+        return deployments
+
+    def get_deployment(self, id_token):
+        deployments = self.__class__.get_deployments(
+            self.user,
+            id_token,
+            search_name=self.chart_name,
+            # search_version=tool_deployment.tool.version,
+        )
```

Comment:
Having `get_deployment` as an instance method feels a bit weird. Maybe this method should be renamed something like `load(self, id_token)` and store the result in `self._deployment`?

Reply:
I'm not that keen on state for this as it mud the water, we wouldn't use this in other methods anyway and if we do it also means that some client code may rely on this method being called before something else (and this instance attribute being set) which may lead to bugs.

As this is a refactoring PR focused on moving these function in the class I'm not going to change anything else - including the use of `release_name` which *may* change the actual behaviour. We can change that independently.